gems/pending/VolumeManager/MiqLvm.rb

```diff
@@ -295,6 +302,436 @@ def parseString
+  BITMAP_HEADER = BinaryStruct.new([
+    'L',                      'csum',
+    'L',                      'notused',
+    'Q',                      'blocknr'
+  ])
+
+  DEVICE_DETAILS = BinaryStruct.new([
+    'Q',                      'mapped_blocks',
+    'Q',                      'transaction_id',
+    'L',                      'creation_time',
+    'L',                      'snapshotted_time'
+  ])
+
+  MAPPING_DETAILS = BinaryStruct.new([
+    'Q',                       'value'
+  ])
+
+  class BTree
+    FLAGS = { :internal => 1, :leaf => 2}
+
+    attr_accessor :root_address
+
+    def initialize(superblock, root_address, value_type)
+      @superblock   = superblock
+      @root_address = root_address
+      @value_type   = value_type
+    end
+
+    def root
+      @root ||= begin
+        @superblock.seek root_address
+        @superblock.read_struct DISK_NODE
+      end
+    end
+
+    def internal?
+      (root['flags'] & FLAGS[:internal]) != 0
+    end
+
+    def leaf?
+      (root['flags'] & FLAGS[:leaf]) != 0
+    end
+
+    def num_entries
+      @num_entries ||= root['nr_entries']
+    end
+
+    def max_entries
+      @max_entries ||= root['max_entries']
+    end
+
+    def key_address(i)
+      root_address + DISK_NODE.size + i * 8
+    end
+
+    def value_base
+      key_address(max_entries)
+    end
+
+    def value_address(i)
+      value_base + @value_type.size * i
+    end
+
+    def keys
+      @keys ||=
+        0.upto(num_entries - 1).collect do |i|
+          @superblock.seek key_address(i)
+          @superblock.read_64.unpack('Q').first
+        end
+    end
+
+    def entries
+      @entries ||=
+        0.upto(num_entries - 1).collect do |i|
+          @superblock.seek value_address(i)
+          @superblock.read_struct @value_type
+        end
+    end
+
+    def entry_for(key)
+      entries[keys.index(key)]
+    end
+
+    def to_h
+      @h ||=
+        Hash[0.upto(num_entries-1).collect do |i|
+          k = keys[i]
+          e = entries[i].kind_of?(BTree) ? entries[i].to_h : entries[i]
+          [k, e]
+        end]
+    end
+
+    def [](key)
+      return to_h[key]
+    end
+  end
+
+  class DataMap < BTree
+    TIME_MASK = (1 << 24) - 1
+
+    def initialize(superblock, root_address)
+      super superblock, root_address, MAPPING_DETAILS
+    end
+
+    alias :device_blocks :keys
+
+    def entries
+      @dmentries ||= begin
+        super.collect do |entry|
+          value = entry['value']
+          internal? ? DataMap.new(@superblock, @superblock.md_block_address(value)) :
+                      [extract_data_block(value), extract_time(value)]
+        end
+      end
+    end
+
+    def data_block(device_block)
+      device_blocks.reverse.each do |map_device_block|
+        if map_device_block <= device_block
+          entry = entry_for(map_device_block)
+          return entry.data_block(device_block) if entry.is_a?(DataMap)
+          raise RuntimeError unless map_device_block == device_block
```

Comment:
Can you raise a specific error message...a caller that gets here will have no idea why this failed.